test/server/admin/stats_handler_test.cc

```diff
@@ -133,6 +135,1420 @@ TEST_P(AdminStatsTest, HandlerStatsPlainText) {
+    "stats": [
+        {
+            "name":"c1",
+            "value":10,
+        }
+    ]
+})EOF";
+
+  Http::Code code = handler.handlerStats(url, response_headers, data, admin_stream);
+  EXPECT_EQ(Http::Code::OK, code);
+
+  EXPECT_THAT(expected_json, JsonStringEq(data.toString()));
+
+  code = handler.handlerStats(url + "&histogram_buckets=cumulative", cumulative_response_headers,
+                              cumulative_data, admin_stream);
+  EXPECT_EQ(Http::Code::OK, code);
+
+  EXPECT_THAT(expected_json, JsonStringEq(cumulative_data.toString()));
+
+  code = handler.handlerStats(url + "&histogram_buckets=disjoint", disjoint_response_headers,
+                              disjoint_data, admin_stream);
+  EXPECT_EQ(Http::Code::OK, code);
+
+  EXPECT_THAT(expected_json, JsonStringEq(disjoint_data.toString()));
+
+  shutdownThreading();
+}
+
+TEST_P(AdminStatsTest, HandlerStatsJsonHistogramBucketsCumulative) {
+  const std::string url = "/stats?histogram_buckets=cumulative&format=json";
+  Http::TestResponseHeaderMapImpl response_headers, used_response_headers,
+      used_and_filter_response_headers;
+  Buffer::OwnedImpl data, used_data, used_and_filter_data;
+  MockAdminStream admin_stream;
+  Configuration::MockStatsConfig stats_config;
+  EXPECT_CALL(stats_config, flushOnAdmin()).WillRepeatedly(testing::Return(false));
+  MockInstance instance;
+  store_->initializeThreading(main_thread_dispatcher_, tls_);
+  EXPECT_CALL(instance, stats()).WillRepeatedly(testing::ReturnRef(*store_));
+  EXPECT_CALL(instance, statsConfig()).WillRepeatedly(testing::ReturnRef(stats_config));
+  StatsHandler handler(instance);
+
+  store_->counterFromString("c1");
+  Stats::Counter& c2 = store_->counterFromString("c2");
+
+  c2.add(20);
+
+  Stats::Histogram& h1 = store_->histogramFromString("h1", Stats::Histogram::Unit::Unspecified);
+
+  EXPECT_CALL(sink_, onHistogramComplete(Ref(h1), 200));
+  h1.recordValue(200);
+
+  EXPECT_CALL(sink_, onHistogramComplete(Ref(h1), 300));
+  h1.recordValue(300);
+
+  store_->mergeHistograms([]() -> void {});
+
+  EXPECT_CALL(sink_, onHistogramComplete(Ref(h1), 300));
+  h1.recordValue(300);
+
+  store_->mergeHistograms([]() -> void {});
+
+  Http::Code code = handler.handlerStats(url, response_headers, data, admin_stream);
+  EXPECT_EQ(Http::Code::OK, code);
+
+  const std::string expected_json = R"EOF({
+    "stats": [
+        {
+            "name":"c1",
+            "value":0
+        },
+        {
+            "name":"c2",
+            "value":20
+        },
+        {
+            "histograms": [
+                {
+                    "name": "h1",
+                    "supported_buckets": [
+                        0.5,
+                        1,
+                        5,
+                        10,
+                        25,
+                        50,
+                        100,
+                        250,
+                        500,
+                        1000,
```

Comment:
it would be nice if there were a way to reduce the number of buckets for the purpose of this test, just so the output is more human readable. It's nicely indented of course but it's way too long :)

Reply:
Thanks for #19928!

I changed the amount of buckets for some tests to clean it up a bit more.

@jmarantz I was also previously thinking about something related to this. Would it make sense to have a configurable default set of buckets for histograms? Currently the default buckets are set for histograms if nothing is found in the config. In #19180 I added a histogram that would only be configured with a certain set of buckets. Could this be implemented directly into Envoy without having to configure it in the config?

Depending on how this would be implemented, something like this could work here too: `histogramFromString("h1", Stats::Histogram::Unit::Unspecified, {1,2,3,4});` where {1,2,3,4} would be the buckets that are given to this specific histogram.